### PR TITLE
CUDA fixes for MCMC

### PR DESCRIPTION
**Pull request by neerajprad**
This PR has some last remaining fixes for CUDA:

 - [x] I was seeing a segmentation fault towards the very end of the baseball example, but all reported results were correct. gdb showed the error coming from this [file](https://github.com/pytorch/pytorch/blob/master/torch/csrc/CudaIPCTypes.cpp). It seems that this is related to a recent PyTorch change that implements reference counting for shared CUDA tensors (https://github.com/pytorch/pytorch/pull/16854). Since we were terminating our workers before the tensors were stacked, it resulted in a segmentation fault during deallocation. This is now fixed by only terminating workers once the tensors are stacked in the main process.
 - [x] ~With https://github.com/pytorch/pytorch/pull/16854, we also don't need to use `mp.Event` anymore and this has been removed. This makes the code slightly faster for small models.~ This isn't true - see discussion below.
 - [x] Moved to using the default initialization as in Stan and NumPyro, in `initialize_model`. This isn't exposed to the user currently since it might be worth refactoring this part to expose multiple initialization strategies as in NumPyro (https://github.com/pyro-ppl/numpyro/pull/237).
 - [x] Checking for both the finiteness of potential energy and its gradient when computing valid initial parameters.
 - [x] Resolving other minor CUDA test failures by setting `mp_context` to `spawn` for CUDA tensors as per PyTorch guidelines.
 - [x] Check that all remaining MCMC tests pass on CUDA.

### CODE REVIEW

File: pyro/infer/mcmc/api.py [hunk 1]
```diff
@@ -231,8 +234,9 @@ def run(self, *args, **kwargs):
                     yield val, chain_id
                 else:
                     active_workers -= 1
+            exc_raised = False
         finally:
-            self.terminate()
+            self.terminate(terminate_workers=exc_raised)
 
 
 class MCMC(object):
```
Comment (fehiepsi) [inline pyro/infer/mcmc/api.py]: could you update the docs for `hook_fn`?

File: pyro/infer/mcmc/util.py [hunk 2]
```diff
@@ -281,21 +285,32 @@ def get_potential_fn(self, jit_compile=False, skip_jit_warnings=True, jit_option
...
+            if strategy == "uniform":
+                params = {k: dist.Uniform(v.new_full(v.shape, -2), v.new_full(v.shape, 2)).sample()
+                          for k, v in params.items()}
+            elif strategy == "prior":
+                trace = poutine.trace(model).get_trace(*model_args, **model_kwargs)
+                samples = {name: trace.nodes[name]["value"].detach() for name in params}
+                params = {k: transforms[k](v) for k, v in samples.items()}
+            pe_grad, pe = potential_grad(potential_fn, params)
+
+            if torch.isfinite(pe) and all(map(torch.all, map(torch.isfinite, pe_grad.values()))):
```
Comment (fehiepsi) [inline pyro/infer/mcmc/util.py]: +1

File: pyro/infer/mcmc/api.py (hunk 1 above)
Comment (neerajprad) [reply pyro/infer/mcmc/api.py]: Oops..thanks for catching this.

**Comment by neerajprad** [comment]
> I'll need to double check CUDA issue in my system because the current PyTorch docs suggests to use Event.

I think the docs are just using event to ensure that the worker process doesn't quit until the producer is done with the tensors! But please do check so that we don't end up facing the same issues as earlier regarding receiving garbage values.

**Comment by fehiepsi** [comment]
@neerajprad The example runs smoothly in my system. Nicely clean up!

**Comment by neerajprad** [comment]
> I think the docs are just using event to ensure that the worker process doesn't quit until the producer is done with the tensors

@fehiepsi - I think we need to still use `events` actually. It does seem that the tensors can be gc'd while they are in the queue if the producer process doesn't hold a reference to it. I ran your example from https://github.com/pytorch/pytorch/issues/16141#issuecomment-455893253 and it fails without using events, so I think we'll need to revert that change. I think the reason why it seems to work in our case is that the consumer is faster than the producers.

File: tests/infer/mcmc/test_nuts.py [hunk 3]
```diff
@@ -184,7 +184,9 @@ def model(data):
     assert_equal(samples["p_latent"].mean(0), true_probs, prec=0.02)
 
 
-@pytest.mark.parametrize("jit", [False, mark_jit(True)], ids=jit_idfn)
+@pytest.mark.parametrize("jit", [False, mark_jit(True, marks=[
+    pytest.mark.skipif("CUDA_TEST" in os.environ, reason="FIXME: Too small step size with CUDA on JIT.")
+])], ids=jit_idfn)
```
Comment (fehiepsi) [inline tests/infer/mcmc/test_nuts.py]: It is tricky to debug the root of this. All I can tell is potential_fn gives wrong derivatives under CUDA+JIT for this type of model. Gamma prior does not cause problem because when I replace `Gamma` prior by `HalfNormal(rate)`, I still get the grad issue. This looks weird to me because this issue does not happen for some other models. I think we need to create an issue to track this problem.

File: pyro/infer/mcmc/api.py [hunk 4]
```diff
@@ -371,11 +388,13 @@ def get_samples(self, num_samples=None, group_by_chain=False):
...
-                batch_dim = 0
-            else:
-                batch_dim = 1
+            batch_dim = 0
+            if self.num_chains > 1:
+                if group_by_chain:
+                    batch_dim = 1
+                else:
+                    samples = {k: v.reshape((-1,) + v.shape[2:]) for k, v in samples.items()}
+                    batch_dim = 0
```
Comment (fehiepsi) [inline pyro/infer/mcmc/api.py]: haha, the logic here is tricky to detect :D

File: tests/infer/mcmc/test_nuts.py (hunk 3 above)
Comment (neerajprad) [reply tests/infer/mcmc/test_nuts.py]: Thanks for digging into this!
